src/main/java/org/truffleruby/core/array/ArrayNodes.java

```diff
@@ -1581,6 +1581,62 @@ public DynamicObject replace(DynamicObject array, DynamicObject other,
 
     }
 
+    @Primitive(name = "array_rotate_inplace", needsSelf = false, lowerFixnum = 2)
+    @ImportStatic(ArrayGuards.class)
+    public abstract static class RotateInplaceNode extends PrimitiveArrayArgumentsNode {
+
+        @Specialization(guards = "strategy.matches(array)", limit = "ARRAY_STRATEGIES")
+        public DynamicObject rotate(DynamicObject array, int amount,
+                @Cached("of(array)") ArrayStrategy strategy) {
+            final int size = strategy.getSize(array);
+            assert 0 < amount && amount < size;
+            final ArrayMirror mirror = strategy.newMirror(array);
+
+            rotateReverse(amount, size, mirror);
+
+            return array;
+        }
+
+        protected void rotateReverse(int amount, int size, ArrayMirror mirror) {
+            final int last = size - 1;
+            reverse(mirror, amount, last);
+            reverse(mirror, 0, amount - 1);
+            reverse(mirror, 0, last);
```

Comment:
I think these 3 phases of rotating could benefit from comments on which portion you're rotating. I wasn't sure why you needed this last one until I realized it was to correct the order due to the previous two leaving the rotated array in reverse order.